### PR TITLE
ARROW-236: Bridging IO interfaces under the hood in pyarrow

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

File: cpp/src/arrow/parquet/reader.h [hunk 1]
```diff
@@ -99,7 +106,7 @@ class ARROW_EXPORT FileReader {
   virtual ~FileReader();
 
  private:
-  class Impl;
+  class PARQUET_NO_EXPORT Impl;
```
Comment (xhochy) [inline cpp/src/arrow/parquet/reader.h]: This should be `ARROW_NO_EXPORT`?

File: cpp/src/arrow/parquet/writer.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ class ARROW_EXPORT FileWriter {
   MemoryPool* memory_pool() const;
 
  private:
-  class Impl;
+  class PARQUET_NO_EXPORT Impl;
```
Comment (xhochy) [inline cpp/src/arrow/parquet/writer.h]: See above

Comment (wesm) [inline cpp/src/arrow/parquet/writer.h]: Oops, yes, thanks

**Comment by wesm** [comment]
Miraculously, I was able to get this working mere minutes before my talk at Data Science Summit. Let me know any comments on the general approach.

**Comment by xhochy** [comment]
The general approach looks good, +1.

**Comment by wesm** [comment]
Thanks -- I'll get the build passing and merge. I am not able to read very many flat Parquet files right now (for example: Impala's Parquet files do not have the UTF8 annotation for strings, similarly timestamps are stored in Int96), so will create a bunch of JIRAs to track these.

In the absence of a separate metadata (like the Hive metastore), we'll have to make some default guesses about the actual schema and eventually provide some options to set the column logical types explicitly when there is ambiguity.

**Comment by wesm** [comment]
+1

**Pull request closed without merging**